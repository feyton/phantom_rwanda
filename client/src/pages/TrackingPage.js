import L from 'leaflet';
import React, { useEffect, useRef, useState } from 'react';
import {
  LayersControl,
  MapContainer,
  TileLayer,
  ZoomControl
} from 'react-leaflet';
import { toast } from 'react-toastify';
import Button, { ButtonLoading } from '../components/Button.js';
import { axiosBase } from '../utils/Api.js';
import socket from '../utils/socket.js';
import ActiveBuses from './ActiveBuses.js';
import RoutingMachine from './RoutingMachine.js';

const TrackingPage = () => {
  const [options, setoptions] = useState([]);
  const [loading, setloading] = useState(false);
  const [activeRoute, setActiveRoute] = useState();
  const { BaseLayer } = LayersControl;
  const rMachine = useRef();

  const handleRoute = async (e) => {
    e.preventDefault();
    options.filter((option) => {
      const route = option.code === e.target.origin.value;
      if (route) {
        setloading(true);
        if (rMachine.current) {
          socket.emit('leave_room', {
            code: activeRoute.code
          });
          rMachine.current.setWaypoints([
            L.latLng(option.start),
            L.latLng(option.end)
          ]);
          rMachine.current.on('routeselected', () => {
            setloading(false);
          });
        } else {
          setActiveRoute(option);
          setloading(false);
        }
      }
    });
  };

  useEffect(() => {
    const fetchOptions = async () => {
      try {
        setloading(true);
        const res = await axiosBase.get('/simulate/options');
        setoptions(res.data.data.options);
      } catch (error) {
        toast('Something went wrong on the server', { type: 'error' });
      } finally {
        setloading(false);
      }
    };
    fetchOptions();
  }, []);

  useEffect(() => {
    if (activeRoute) {
      socket.emit('join_room', {
        code: activeRoute.code
      });
    }
  }, [activeRoute]);

  return (
    <div className="flex flex-col">
      <form
        id="form"
        onSubmit={handleRoute}
        className="fixed z-20 flex flex-col justify-center"
      >
        <div className="flex flex-col bg-background p-4 rounded-xl mt-2 ml-6 md:ml-10 ">
          <select
            type="text"
            id="origin"
            name="origin"
            placeholder="your current location"
            className="rounded-xl bg-background border-primary text-sm outline-none mb-2 pl-4 md:pl-8 py-1  w-40 md:w-56 border-2"
          >
            <option id="origin-select">Select route</option>
            {options.map((option) => {
              return (
                <option
                  value={option.code}
                  key={option.code}
                  className="cursor-pointer bg-transparent font-bold font-raleway disabled:text-gray-400 disabled:bg-gray-100"
                >
                  {`${option.origin} - ${option.destination}`}
                </option>
              );
            })}
          </select>
          <div type="submit" className="flex justify-center">
            {loading ? (
              <ButtonLoading name="Loading..." />
            ) : (
              <Button
                name="Track"
                styles="bg-primary text-sm text-background py-1"
              />
            )}
          </div>
        </div>
      </form>
      <div className="flex flex-col items-center z-10">
        <MapContainer
          center={{ lat: -1.936671, lng: 30.053524 }}
          zoom={13}
          zoomControl={false}
          className="h-screen md:h-[88vh] w-[95vw]"
        >
          <LayersControl>
            <BaseLayer checked name="Open Street">
              <TileLayer
                attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
                url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
              />
            </BaseLayer>
            <BaseLayer name="Google Sattelite">
              <TileLayer
                url="https://{s}.google.com/vt/lyrs=s&x={x}&y={y}&z={z}"
                maxZoom={20}
                subdomains={['mt1', 'mt2', 'mt3']}
              />
            </BaseLayer>
          </LayersControl>

          <ActiveBuses />
          <ZoomControl position="topright" />

          {options && activeRoute ? (
            <RoutingMachine
              ref={rMachine}
              origin={activeRoute.start}
              destination={activeRoute.end}
            />
          ) : (
            ''
          )}
        </MapContainer>
      </div>
    </div>
  );
};

export default TrackingPage;

// TODO To calculate the use distance
// if (userLocation && buses) {
//   let distances = [];
//   let loc = JSON.parse(userLocation);
//   buses.slice().filter(item=>item.route === activeRoute.code)
//     .forEach((bus) => {
//       const origin = [loc.lat, loc.lng];
//       const dest = [bus.location.latitude, bus.location.longitude];
//       const distance = getDistance(origin, dest);
//       distances.push({ id: bus.entityId, distance, bus });
//     });
//   const nearBus = distances
//     .sort((a, b) => {
//       return a.distance - b.distance;
//     })
//     .filter((item, index, array) => item.distance === array[0].distance)[0];
//   console.log(nearBus);
// }
