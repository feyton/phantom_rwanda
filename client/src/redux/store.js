import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './reducers/authReducer.js';
import { busMovementReducer } from './reducers/busMovementReducer.js';
import { counterReducer } from './reducers/counterReducer.js';
import { dashboardReducer } from './reducers/dashboardReducer.js';
import { managementReducer } from './reducers/managementReducer.js';
import { profileReducer } from './reducers/profileReducer.js';
import { simulateReducer } from './reducers/simulateReducer.js';
import { socketReducer } from './reducers/socketReducer.js';
import { updateProfileReducer } from './reducers/updateProfileReducer.js';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    profileUpdates: updateProfileReducer,
    profileInfo: profileReducer,
    auth: authReducer,
    management: managementReducer,
    dashboard: dashboardReducer,
    socket: socketReducer,
    simulate: simulateReducer,
    movement: busMovementReducer
  }
});
