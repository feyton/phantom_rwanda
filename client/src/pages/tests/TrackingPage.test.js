import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { store } from '../../redux/store.js';
import TrackingPage from '../TrackingPage.js';

describe('TrackingPage', () => {
  it('should render TrackingPage', () => {
    const elem = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <TrackingPage />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

  it('It should test the handleRoute function', () => {
    const handleRoute = jest.fn();
    const event = {
      target: {
        origin: { value: 'some value' },
      }
    };
    const component = mount(
      <MemoryRouter>
        <Provider store={store}>
          <TrackingPage handleRoute={handleRoute()} />
        </Provider>
      </MemoryRouter>
    );
    const form = component.find('form');
    form.simulate('submit', event);
    expect(handleRoute).toBeCalledTimes(1);
  });

  it('It should test the handleChange function', () => {
    const handleChange = jest.fn();
    const component = mount(
      <MemoryRouter>
        <Provider store={store}>
          <TrackingPage handleChange={handleChange()} />
        </Provider>
      </MemoryRouter>
    );
    const form = component.find('#origin');
    form.simulate('change');
    expect(handleChange).toBeCalledTimes(1);
  });

});

describe('Tracking page', () => {


  it('should render the tracking page', () => {
    mount(
      <MemoryRouter>
        <Provider store={store}>
          <TrackingPage />
        </Provider>
      </MemoryRouter>
    );
  });
});

