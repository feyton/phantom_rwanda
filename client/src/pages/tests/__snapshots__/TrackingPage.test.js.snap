// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackingPage should render TrackingPage 1`] = `
<div
  className="flex flex-col"
>
  <form
    className="fixed z-20 flex flex-col justify-center"
    id="form"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col bg-background p-4 rounded-xl mt-2 ml-6 md:ml-10 md:items-baseline"
    >
      <select
        className="rounded-xl bg-background border-primary text-sm outline-none mb-2 pl-4 md:pl-8 py-1  w-40 md:w-56 border-2"
        id="origin"
        name="origin"
        placeholder="your current location"
        type="text"
      >
        <option
          id="origin-select"
        >
          Select route
        </option>
      </select>
      <div
        className="flex justify-center md:ml-16"
        type="submit"
      >
        <div>
          <button
            className="bg-primary text-sm text-background py-1 rounded-xl py-2 text-base md:text-sm px-5"
            type="submit"
          >
            Track
          </button>
        </div>
      </div>
    </div>
  </form>
  <div
    className="flex flex-col items-center z-10"
  >
    <div
      className="h-screen md:h-[88vh] w-[95vw]"
    />
  </div>
</div>
`;
